### PR TITLE
Add Ketcher KET (JSON) chemical structure format

### PR DESCRIPTION
**Pull request by romanbsd**
## Summary

- Adds a new read/write file format plugin for [KET](https://github.com/epam/Indigo), Ketcher's native JSON chemical-structure format (also produced and consumed by Indigo). Registered under the `ket` ID; lives in `src/formats/json/ketformat.cpp`.
- Supports molecules (atoms, bonds, S-groups, R-group attachment sites, fragment-level properties, highlights) and reactions (arrow + plus + molecule layout → `OBReactionFacade`).
- Sections outside OpenBabel's native data model (macromolecule monomers, monomer/ambiguous/group templates, V2 rich text, images, simple shapes, monomer shapes, `root.connections`, `root.annotation`, `ket_version`) are preserved verbatim across round-trips so Ketcher-only data is not lost.

## Round-trip fidelity

- Preserves original `$ref` names (e.g. `mol5`, not renamed to `mol0`) and the exact `root.nodes` ordering, including interleaved inline meta objects (arrows, pluses, text, images, simpleObjects, multi-tailed-arrow).
- Preserves `root.connections` verbatim — important for macromolecule backbones that the OpenBabel data model can't natively represent.
- Original reaction arrows and pluses are kept on output rather than synthesized from centroids; the writer only synthesizes them when `OBMol::IsReaction()` is true AND no original of that type exists in the preserved node order.
- S-group atom and bond indices are remapped back to per-molecule-local KET indices on write.
- Per-molecule passthrough fields (highlight, properties, stereoFlagPosition) are stored on the first atom of each component so they don't leak across components in a multi-mol document.
- `ket_version` is emitted whenever the input's major version exceeds 1.

## Test plan

- [x] Build: `cmake .. -DCMAKE_POLICY_VERSION_MINIMUM=3.5 -DCMAKE_BUILD_TYPE=Debug -DCMAKE_FIND_FRAMEWORK=LAST -DRUBY_BINDINGS=OFF && make -j8`
- [x] `ctest -R ketformat` — 13/13 pass (parts 1–13 of `ketformattest.cpp`).
- [x] `obabel -iket sample.ket -oket -xm` end-to-end: verbatim preservation of `monomer`, `monomerTemplate-*`, `image`, `text`, `annotation`, `connections`.
- [x] Python test (`test/testketformat.py`) covers read, write, byte-stable double round-trip, reaction assembly, per-component scoping of highlight/properties, passthrough preservation. Registered under `pybindtests`.
- [ ] CI on this PR

## Test fixtures (`test/files/ket/`)

| Fixture | Exercises |
|---|---|
| `ethanol.ket` | minimal single molecule |
| `multi.ket` | two components with formal charges |
| `reaction.ket` | reactants + arrow + plus + product |
| `with_passthrough.ket` | `ket_version: 2.0.0` + monomer + monomerTemplate + image |
| `with_connections.ket` | `root.connections`, non-default `\$ref` (`mol5`), node order (text between mols), template, annotation |
| `full_fidelity.ket` | multi-mol with S-groups (GEN + DAT, with bonds), highlight, properties |

## Notes

- The implementation deliberately does not attempt to expand macromolecule monomers / monomer templates / R-group fragments into atoms. Those payloads are preserved verbatim — translating them would require structural changes outside this format plugin's scope.
- Bond types 5–10 (query / coordination / hydrogen) round-trip via a preserved `_ket_bond_type` OBPairData attribute; the OBMol bond order is set to 1 in those cases.
- Reaction roles are assigned by projecting each component's centroid onto the arrow axis. Components whose perpendicular distance from the axis exceeds half the arrow length are tagged as agents.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added support for reading and writing Ketcher KET (JSON) format files, including molecules, multi-component structures, and reactions with full structural fidelity preservation.
  * Supports minified output options for compact file storage.
  * Preserves metadata including charge information, atom properties, and bond types.

* **Tests**
  * Added comprehensive test suite with multiple test fixtures covering molecule and reaction parsing, round-trip validation, structural metadata preservation, and edge cases.

[Review Change Stack](https://app.coderabbit.ai/change-stack/openbabel/openbabel/pull/2909)

### CODE REVIEW

**Comment by ghutchis** [comment]
Was this entirely generated with Claude Code? While we have not adopted an AI use policy for Open Babel, I think it's critical to ensure everything is fully reviewed by a human.

For example for Avogadro, we have adapted:
https://avogadro.cc/contrib/ai-usage.html

I'm very happy to support Ketcher format - but I'd like to know a bit more about the level of automation here.

**Comment by romanbsd** [comment]
It was reviewed by me and Gemini and Codex as well.

--Roman

File: src/formats/json/ketformat.cpp [hunk 1]
```diff
@@ -0,0 +1,1833 @@
...
+
+    bool parse(const rapidjson::Document &doc, std::string &errorMsg)
+    {
+        if (!doc.IsObject() || !doc.HasMember("root") || !doc["root"].IsObject()) {
+            errorMsg = "KET document must have a 'root' object";
+            return false;
+        }
+
+        if (doc.HasMember("ket_version") && doc["ket_version"].IsString())
+            setPairData(mol_, kAttrKetVersion, doc["ket_version"].GetString());
```
Comment (ghutchis) [inline src/formats/json/ketformat.cpp]: Is there any worry about compatibility - should it be checking the ket_version to make sure this implementation will work?

File: src/formats/json/ketformat.cpp [hunk 2]
```diff
@@ -0,0 +1,1833 @@
...
+        // Aromatic bonds: kekulize if any were flagged.
+        if (needsKekulization_) {
+            mol_->SetAromaticPerceived();
+            FOR_BONDS_OF_MOL(b, mol_) {
+                if (b->IsAromatic()) {
+                    b->GetBeginAtom()->SetAromatic();
+                    b->GetEndAtom()->SetAromatic();
+                }
+            }
+            OBKekulize(mol_);
```
Comment (ghutchis) [inline src/formats/json/ketformat.cpp]: Do you want Open Babel to do Kekule perception? I'd think you'd want to either leave as fully aromatic or fully Kekule depending on how the sketch handled it.

File: src/formats/json/ketformat.cpp [hunk 3]
```diff
@@ -0,0 +1,1833 @@
...
+                    std::string(buf.GetString(), buf.GetSize()));
+    }
+
+    void parseRGroup(const std::string &refName, const rapidjson::Value &node)
+    {
+        // R-group payloads are full molecule fragments. We store them as
+        // passthrough so the writer can echo them — translating them into
+        // OpenBabel's mol-of-fragments model would require structural
+        // changes outside this format's scope.
+        passthrough_.unknownRefs.emplace_back(refName, stringifyValue(node));
```
Comment (ghutchis) [inline src/formats/json/ketformat.cpp:976]: Maybe, but sometimes people want to expand the R group aliases.
https://openbabel.org/docs/Command-line_tools/babel.html#aliases-for-chemical-groups

**Review by ghutchis — CHANGES_REQUESTED** [review]
At the moment, I'm just skimming. (I also think you need to merge from HEAD) but there are a few obvious things.

I also think the common conversions will include CDXML and SDF and those seem like they could use specific unit tests?

**Comment by ghutchis** [comment]
@coderabbitai review

File: src/formats/json/ketformat.cpp (hunk 2 above)
Comment (romanbsd) [reply src/formats/json/ketformat.cpp]: I'll leave them

File: src/formats/json/ketformat.cpp [hunk 4]
```diff
@@ -0,0 +1,1833 @@
...
+    if (getPairDataString(atom, kAttrKetUnsaturatedAtom, s) && s == "true")
+        obj.AddMember("unsaturatedAtom", true, al);
+    if (getPairDataInt(atom, kAttrKetHCount, v))
+        obj.AddMember("hCount", v, al);
+
+    // Emit implicitHCount if it differs from OB's default H perception.
+    if (atom->GetImplicitHCount() > 0) {
+        obj.AddMember("implicitHCount",
+                      static_cast<int>(atom->GetImplicitHCount()), al);
+    }
```
[the remark this answers is not in the document]
Comment (ghutchis) [reply src/formats/json/ketformat.cpp]: @romanbsd - any thoughts on this one? Seems like if it's explicitly set to zero in the file, you'd want that.

**Comment by ghutchis** [comment]
See my comment above, but otherwise, this is in good shape.

**Comment by romanbsd** [comment]
Yeah, I'll give it another pass probably today or tomorrow. We're also testing it internally with different molecules.

File: src/formats/json/ketformat.cpp (hunk 1 above)
Comment (romanbsd) [reply src/formats/json/ketformat.cpp]: Good point. I added a compatibility check for `ket_version`: unsupported future major versions are rejected explicitly (currently major > 2), and there is test coverage for 3.0.0 rejection. Version 2.x payloads are still accepted and preserved for round-trip fidelity.

File: src/formats/json/ketformat.cpp (hunk 3 above)
Comment (romanbsd) [reply src/formats/json/ketformat.cpp:976]: Handled by keeping the default behavior fidelity-preserving while adding an opt-in path for expansion. By default KET aliases are exposed as `AliasData` and round-trip as aliases; with the `-ia` read option, chemically meaningful aliases such as `COOH` are expanded using OpenBabel's alias table. I added C++ coverage for the `-ia` expansion path and matching Python binding coverage.

File: src/formats/json/ketformat.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (romanbsd) [reply src/formats/json/ketformat.cpp]: this was fixed

**Comment by romanbsd** [comment]
I think that everything was addressed. Do you want me to squash commits?

**Comment by ghutchis** [comment]
I'll squash. I need to figure out why a fuzz test is failing on your branch but okay on master. I'll pull it down and investigate.

**Comment by ghutchis** [comment]
Okay, that should do it - as far as I can tell, the problem was in a SWIG test because the Linux Python was freeing a molecule fast (but not happening on my Mac)

Assuming this builds clean, I'll be happy to merge

**Review by ghutchis — APPROVED** [review]
LGTM, thanks
